### PR TITLE
GRDB.swift  implementation - Tsungyu Yu

### PR DESCRIPTION
**Pull request by ytyubox**
Hi Caio and Mike,

This is my pull request for Feed Store Challenge, my second attempt in GRDB.swift.
Please have a look, thanks. 😁

### CODE REVIEW

File: FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+//
+//  Created by 游宗諭 on 2020/12/3.
+//  Copyright © 2020 Essential Developer. All rights reserved.
+//
+
+import GRDB
+public protocol DBQueueManager {
+    func makeQueue(path: String?) throws -> DatabaseQueue
+    func recylcle(queue: DatabaseQueue)
+}
```
Comment (caiozullo) [inline FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift:13]: This abstraction seems unnecessary. You could remove it.

File: FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+public class DBQueueFactory: DBQueueManager {
+    public init() {}
+    private var pool:[String?: (count:Int,dbQueue: DatabaseQueue)] = [:]
+    public func makeQueue(path: String?) throws -> DatabaseQueue {
```
Comment (caiozullo) [inline FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift:18]: Can the `path` be nil? How would it work?

Moreover, the `path` is being used to create the `DatabaseQueue`, so it may not work as intended.

File: Tests/FeedStoreIntegrationTests.swift [hunk 3]
```diff
@@ -6,81 +6,111 @@ import XCTest
...
+        let storeToInsert = makeSUT(dbQueueManager: manager)
+        let storeToDelete = makeSUT(dbQueueManager: manager)
+        let storeToLoad = makeSUT(dbQueueManager: manager)
+
+        insert((uniqueImageFeed(), Date()), to: storeToInsert)
+
+        deleteCache(from: storeToDelete)
+
+        expect(storeToLoad, toRetrieve: .empty)
+    }
```
Comment (caiozullo) [inline Tests/FeedStoreIntegrationTests.swift:72]: To pass the challenge, the tests shouldn't be modified. Revert the change in the tests and implement only the `makeSUT` function.

File: FeedStoreChallenge/GRDBFeedStore/GRDBFeedStore.swift [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    
+    public init(path: String?, dbManeger: DBQueueManager) throws {
+        self.dbManeger = dbManeger
+        dbQueue = try dbManeger.makeQueue(path: path)
+    }
+    deinit {
+        dbManeger.recylcle(queue: dbQueue)
+    }
+    public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+        try? dbQueue.write({ (db) in
```
Comment (caiozullo) [inline FeedStoreChallenge/GRDBFeedStore/GRDBFeedStore.swift:27]: Errors should be handled. But `try?` will ignore errors.

File: FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift (hunk 2 above)
Comment (ytyubox) [reply FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift:14]: In the library, if we want to use a in memory database, we can use the hard code path: `:memory:`. 
And yes, I did not use the path for my DatabaseQueue, and its fix by next commit, thank you.

File: FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift (hunk 1 above)
Comment (ytyubox) [reply FeedStoreChallenge/GRDBFeedStore/GRDBFeedDataBase.swift]: Yes, totally agree.

File: Tests/FeedStoreIntegrationTests.swift (hunk 3 above)
Comment (ytyubox) [reply Tests/FeedStoreIntegrationTests.swift]: The DBQueueFactory now become a Static member (let) of GRDBFeedStore.

File: FeedStoreChallenge/GRDBFeedStore/GRDBFeedStore.swift (hunk 4 above)
Comment (ytyubox) [reply FeedStoreChallenge/GRDBFeedStore/GRDBFeedStore.swift:26]: Sure, the issue has been fixed by follow up commit.

**Review by caiozullo — COMMENTED** [review]
Hi @ytyubox,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅ 
- You broke down the implementation into small commits ✅ 
- Bonus points for the Integration Tests and Memory Leak Tracking ✅ 

Great job!
